core/src/main/java/io/grpc/transport/MessageDeframer.java

```diff
@@ -174,6 +303,117 @@ public void deframe(ReadableBuffer data, boolean endOfStream) {
   }
 
   /**
+   * Add messages to the delivery queue. This is only called by the transport thread.
+   */
+  private void enqueueFrames() {
+    ArrayDeque<Frame> queue = queueRef.getAndSet(null);
+    if (queue == null) {
+      // Application thread has the queue so create a new one and accumulate into it so we don't
+      // block on the application thread.
+      queue = new ArrayDeque<Frame>();
+    }
+
+    boolean readBody = false;
+    boolean queuedEos = false;
+    long unreturnedBytes = bytesReceived - bytesReturned;
+    long bytesToReturnImmediate = 0;
+    try {
+      // Process the uncompressed bytes.
+      CompositeReadableBuffer buffer = readRequiredBytes();
+      while (buffer != null) {
+        bytesConsumed += buffer.readableBytes();
+        bytesToReturnImmediate += returnBytes(true, queue);
+        switch (state) {
+          case HEADER:
+            processHeader(buffer);
+            break;
+          case BODY:
+            // Read the body and enqueue the message.
+            processBody(buffer, queue);
+            readBody = true;
+            break;
+          default:
+            throw new AssertionError("Invalid state: " + state);
+        }
+        buffer = readRequiredBytes();
+      }
+      if (endOfStream) {
+        if (state == State.BODY || unprocessed.readableBytes() > 0) {
+          // We've received the entire stream and have data available but we don't have
+          // enough to read the next frame ... this is bad.
+          throw Status.INTERNAL.withDescription("Encountered end-of-stream mid-frame")
+              .asRuntimeException();
+        } else {
+          // Can't notify stream yet as there were not enough message requests available
+          // to consume all the messages. Instead we have to enqueue signalling EOS on the
+          // stream.
+          if (transportCount.get() < 0) {
+            queuedEos = true;
+            queue.add(END_OF_STREAM);
+          }
+        }
+      }
+      // Can immediately return all bytes that have not been returned via the queue at this point.
+      bytesToReturnImmediate += returnBytes(false, queue);
```

Comment:
I'm having trouble wrapping my head around the "return bytes" logic.  Correct me if I'm wrong, but it seems that we're always returning the bytes that have been received up to this point.  If that's the case, how does the application push back on flow control?